package.json

```diff
@@ -1,5 +1,6 @@
 {
   "name": "jsbi",
+  "type": "module",
```

Comment:
This is a breaking change. Consumers now have to `import` rather than `require()`. A more incremental step would be to stick to publishing CommonJS. Could we do that?

Reply:
The [docs](https://nodejs.org/api/packages.html#packages_type) seem to indicate this should only have effects on JS files within the JSBI project, and have no bearing on how the package is imported. There's also [instructions](https://nodejs.org/api/packages.html#packages_dual_package_hazard) that indicate how to make a package compatible with both ESM and CJS, and neither of the proposed solutions there indicate that they are a breaking change. We could probably use option 1?

That being said, I did make this change for Temporal, and it caused some issues for certain environments (see https://github.com/js-temporal/temporal-polyfill/issues/29), so if we didn't make that change here I'd revert it there and apply the same solution to both repositories.

We can avoid changing this field by renaming the transpilation output of tsc from .js to .mjs files (by changing the "build" script to be something like `"tsc && for f in tsc-out/*.js; do mv -- \"$f\" \"${f%.js}.mjs\"; done && rollup --config rollup.config.js",`, but it would be nice to avoid that if possible.

Reply:
I think there's a misunderstanding. There are two ways of indicating a package contains JavaScript modules (as opposed to CommonJS):

1. using the `.mjs` extension
2. using `"type": "module"` (which treats all `.js` files as if they were `.mjs`)

You’re right that we could do either of those _if_ we wanted to export JSBI as JavaScript modules (as opposed to the status quo of exporting as CommonJS).

I think it’s just not necessary to make that change, at this point. Why do you want to export as modules? It’s orthogonal to migrating the source to TypeScript.

Ideally, migrating the source to TypeScript is an implementation detail that doesn’t affect what we export or how users should import the npm package.

Reply:
Here’s what we’ve been publishing to npm until now:

```
$ npm pack --dry-run
npm notice 📦  jsbi@3.1.6
npm notice === Tarball Contents ===
npm notice 10.2kB LICENSE
npm notice 27.6kB dist/jsbi-cjs.js
npm notice 32.0kB dist/jsbi-umd.js
npm notice 1.2kB  package.json
npm notice 10.0kB README.md
npm notice 27.6kB dist/jsbi.mjs
npm notice 1.5kB  jsbi.d.ts
```

Ideally, after the TS migration, we still publish the same files with equivalent contents. (It would be nice to only publish the module version in the future, I agree, but it’s a separate change that shouldn’t block the TS migration IMHO.)

Reply:
I think I better understand this now, especially after looking into https://github.com/js-temporal/temporal-polyfill/issues/29 and finding one potential fix (see https://github.com/js-temporal/temporal-polyfill/pull/35).

If we specify that this package uses modules, then we'd have to rename the existing non-module files in dist (so jsbi-cjs.js and jsbi-umd.js) to use a `.cjs` file extension, and while some tools can automatically deal with this others (and especially users loading this file by path on disk) might not (I'm not sure how the UMD bundle in particular is used by tools or humans directly).

I'll remove the key for now and work around the issues it causes within the build system, and file an issue to consider swapping this around in some breaking version of JSBI.

Reply:
Done.